apps/voting-disputable/contracts/DisputableVoting.sol

```diff
@@ -0,0 +1,745 @@
+        return representatives[_voter] == _representative;
+    }
+
+    /**
+    * @dev Tell whether a vote can be paused or not
+    * @param vote_ Vote action instance being queried
+    * @return True if the given vote can be paused, false otherwise
+    */
+    function _canPause(Vote storage vote_) internal view returns (bool) {
+        return vote_.disputableStatus == DisputableStatus.Active && vote_.pausedAt == 0;
+    }
+
+    /**
+    * @dev Tell whether a vote is paused or not
+    * @param vote_ Vote action instance being queried
+    * @return True if the given vote is paused, false otherwise
+    */
+    function _isPaused(Vote storage vote_) internal view returns (bool) {
+        return vote_.disputableStatus == DisputableStatus.Paused;
+    }
+
+    /**
+    * @dev Tell whether a vote is paused or cancelled, or not
+    * @param vote_ Vote action instance being queried
+    * @return True if the given vote is paused or cancelled, false otherwise
+    */
+    function _isPausedOrCancelled(Vote storage vote_) internal view returns (bool) {
+        DisputableStatus status = vote_.disputableStatus;
+        return status == DisputableStatus.Paused || status == DisputableStatus.Cancelled;
+    }
+
+    /**
+    * @dev Internal function to check if a voter has already voted on a certain vote. It assumes the queried vote exists.
+    * @return True if the given voter has not voted on the requested vote, false otherwise
+    */
+    function _hasNotVotedYet(Vote storage vote_, address _voter) internal view returns (bool) {
+        return _voteCaster(vote_, _voter) != _voter;
+    }
+
+    /**
+    * @dev Internal function to check if a vote is still open. It assumes the queried vote exists.
+    * @return True if the given vote is open, false otherwise
+    */
+    function _isVoteOpen(Vote storage vote_) internal view returns (bool) {
+        if (_isPausedOrCancelled(vote_)) {
+            return false;
+        }
+
+        return getTimestamp64() < _voteEndDate(vote_) && !vote_.executed;
+    }
+
+    /**
+    * @dev Internal function to check if a vote is within its overrule window. It assumes the queried vote exists.
+    * @return True if the given vote is within its overrule window, false otherwise
+    */
+    function _withinOverruleWindow(Vote storage vote_) internal view returns (bool) {
+        return getTimestamp64() >= _voteEndDate(vote_).sub(vote_.overruleWindow);
+    }
+
+    /**
+    * @dev Internal function to calculate the end date of a vote. It assumes the queried vote exists.
+    */
+    function _voteEndDate(Vote storage vote_) internal view returns (uint64) {
+        uint64 endDate = vote_.startDate.add(voteTime);
+        DisputableStatus status = vote_.disputableStatus;
+
+        if (status == DisputableStatus.Cancelled) {
+            return vote_.pausedAt.add(vote_.pauseDuration);
+        }
+
+        return endDate.add(vote_.pauseDuration);
```

Comment:
We can optimize this a bit :).

But also feels a bit weird to not handle the `Paused` state at all. This is only used with `_withinOverruleWindow()`, so I wonder if it's something we should inline into that to bundle the assumptions together with where that function is applied (e.g. it never makes sense for `_withinOverruleWindow()` to return true in a paused state).

Reply:
Note that it is also being used in `_isVoteOpen`

Reply:
`Paused` is being handled when doing `endDate.add(vote_.pauseDuration)`. Note that the pause duration is being set when the vote is resumed from being paused.

Reply:
Right... so my problem with `endDate.add(vote_.pauseDuration)` is that it doesn't _really_ handled the paused state.

If the vote is paused, I honestly don't know what the end time will be, and it seems like bad practice to report it as the original end time (it most definitely will not be!).

Reply:
Let me know if the proposed enhancement looks better!